File: 560-SubarraySumEqualsK/560-SubarraySumEqualsK.cpp
// Last updated: 4/3/2026, 1:59:40 PM
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        // Time Complexity: O(N) or O(N*logN) depending on map type
        // Space Complexity: O(N)
        unordered_map<int, int> mp;
        int sum = 0, ans = 0;
        mp[0] = 1; // Set count of 0 to 1
        for (int i = 0; i < nums.size(); i++) {
            // add the current element to the prefix sum
            sum += nums[i];
            // calculate the prefix sum
            int target = sum - k;
            // add the occurrence of the prefix sum to answer
            ans += mp[target];
            // store the current prefix sum
            mp[sum]++;
        }
        return ans;
    }
};

File: 560-SubarraySumEqualsK/560-SubarraySumEqualsK.py
# Last updated: 4/3/2026, 1:59:37 PM
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        sum, cnt = 0, 0
        prefixSum = {0 : 1}
        for x in nums:
            sum += x
            rem = sum - k
            if rem in prefixSum:
                cnt += prefixSum[rem]
            prefixSum[sum] = prefixSum.get(sum, 0) + 1

        return cnt